lnst/Recipes/ENRT/ConfigMixins/DevFlowsPinningHWConfigMixin.py

```diff
@@ -0,0 +1,236 @@
+        enables flow steering across test device queues. Generated flows
+        are iterated and pinned to
+        :attr:`lnst.RecipeCommon.Perf.Measurements.BaseFlowMeasurement.Flow.receiver_cpupin`
+        CPU queue. Flow steering "keys" are configurable by overriding the
+        :meth:`.steer_flow_by` method and can be set independently.
+
+        This is useful especially for cases, when user don't want to rely
+        on rx hash algorithm to steer packets to the right CPU queue and
+        want to pin them manually.
+        Basically, just more-granular Receive side scaling. See
+        https://man7.org/linux/man-pages/man8/ethtool.8.html
+
+        Can be "partially" used by pinning just selected flows and keeping
+        the rest unpinned - so they'll be steered by rx hash algorithm just
+        by returning `None` in :meth:`.steer_flow_by` method for unpinned flow.
+
+        This mixin does NOT pin IRQs to
+        :attr:`lnst.RecipeCommon.Perf.Measurements.BaseFlowMeasurement.Flow.receiver_cpupin`
+        CPUs, which is required for optimal performance.
+    """
+
+    def hw_config(self, config: EnrtConfiguration):
+        """
+        Pins flows generated by
+        :meth:`BaseFlowMeasurementGenerator.generate_flow_combinations`
+        to CPU queues by using `ethtool -N ... flow-type...` command.
+
+        Internal configuration is stored in
+        :attr:`EnrtConfiguration.hw_config.flow_steering_rules` attribute.
+        """
+        super().hw_config(config)
+
+        config.hw_config["flow_steering_rules"] = {}
+
+        for combination in self.generate_flow_combinations(config):
+            for flow in combination:
+                if self.steer_flow_by(flow) is None:
+                    continue  # skip unpinned flow
+
+                for steering_nic in self.steer_flow_at(flow):
+                    if steering_nic not in config.hw_config["flow_steering_rules"]:
+                        config.hw_config["flow_steering_rules"][steering_nic] = []
+
+                    config.hw_config["flow_steering_rules"][steering_nic].append(
+                        self._pin_flow(flow, steering_nic)
+                    )
+
+        return config
+
+    def _pin_flow(self, flow: Flow, steering_nic: Device) -> tuple[str, str]:
+        """
+        Method pins :param:`flow` to :any:`Flow.receiver_cpupin`.
+        Pinning is done by `ethtool -N ... flow-type ... action`,
+        not by `queue`. So other actions (by <0 action code) can
+        be specified as well.
+        See https://man7.org/linux/man-pages/man8/ethtool.8.html
+
+        :param flow: flow object to pin
+        :param steering_nic: NIC where the flow should be pinned
+
+        :returns: tuple of rule ID (for deconfig) and description of pinning.
+            That can't be dynamically generated in :meth:`.describe_hw_config`
+            as it woudln't know which flow belongs to which rule ID.
```

Comment:
Can't you save the Flow:ID mapping in the `config.hw_config` and use it like that ?

Reply:
Ok, thanks for clarifying.